Refactoring info collect and saving on files

guilpier-code opened the pull request:
See comments below

flomnes committed:
Rename TimeElapsed -> Benchmarking

flomnes committed:
Rename timeelapsed -> benchmarking

flomnes committed:
Rename info.h -> timeinfo.h

flomnes committed:
Add struct StudyInfo

guilpier-code committed:
study info file for statistics : dump study number of years and number of enabled thermal clusters

guilpier-code committed:
study info file for statistics : dump areas and links count into study-info.txt

guilpier-code committed:
study info file for statistics : dump BC count, UC mode and nb of years in parallel into study-info.txt

guilpier-code committed:
study info file for statistics : separating file content and file writer into 2 different classes

guilpier-code committed:
study info file for statistics : file content class now has a vector of information item

Class StudyInfoContainer now has a vector of StudyInfoItem instead of a data member for each piece of information collected from study.
Class StudyInfoContainer still collects information on study but use the "addItem" simple method to add an information to its vector of information items.
The file writer now merely loops on file container's vector to print each info item.

guilpier-code committed:
study info file for statistics : the file's CSV writer now inherit form a base abstract class

StudyInfoWriter is now a base abstract class and its child StudyInfoCSVwriter is the writer to the CSV file.
Note that StudyInfoWriter is not exactly an interface, as it has data members supposed to be useful in any of its child writer,
especially a file content and a path to a file.

guilpier-code committed:
study info file for statistics : base writer should contain the content to be stored, but nothing else.

Therefore, our CSV writer now contains the file path as a data member.

flomnes committed:
Merge branch develop into feature/statistics-add-study-info

flomnes committed:
Create & call method Application::writeStudyInfos

flomnes committed:
Save optimization-related infos

flomnes committed:
Collect solver version

flomnes committed:
Update src/libs/antares/benchmarking/studyinfo.h

flomnes committed:
Apply suggestions from code review

guilpier-code committed:
Refactor info collect : extract info collector class + simplifying file content addItem function

Sorry, big commit

guilpier-code committed:
Refactor info collect : divide collect and print in 3 files

These file concern :
- Info collecting
- file content
- file writing

guilpier-code committed:
Refactor info collect : first use a time collector (for study load) - here we just save the work

guilpier-code committed:
[skip ci] Refactor info collect : adding a title to time measurement file

guilpier-code commented:
This is a try to remove the duplicate code between : 
- The way we collect and print data from the study and the simulation in **study-info.txt** in one hand
- The way we collect and print durations form the simulation in **time_measurement.txt** in one the other hand

Until now we roughly used, for both files, the following (collaborating) classes : 
- The sources of data : the study and the simulation
- The file content
- The file writer

In one case (**study-info.txt**), the file content class was polluted with some data collecting, so this class makes too many things.
In the other case (**time_measurement.txt**), the collecting of durations (or times if you like) consists in passing the file content to the simulation and  let a **Timer** class fill the file content on the fly as the simulation runs. I guess a file content should not be passed to the simulation.

So the main idea here is to : 
- extract some information collector classes, whose mission is to collect data (as their name suggests)
- make one file content and one file writer classes used for both printed files.

The way we collect information is different depending on the file we print.
That's why you'll find here 3 data collectors : 
- one to collect data from study (nb of areas, nb of links, ...)
- one to collect data from simulation (nb of variables, constraints, ...)
- one to collect durations from simulation.

The 2 last collectors are passed to the simulation and the last does its job with the help of a very simple timer.

For now, 2 ways of collecting **durations** are **coexisting** (so don't be surprised) : 
- the way we did before this PR (for **time_measurement.txt**)
- a way with the help of a duration collector (so we print another time measurement file : **time_measurement_draft.txt**)

The second is meant to replace the first, or much rather is a start for a discussion.
This way of collecting durations is **not complete**, but gives a good idea of what is looked for.
So, in the code, you'll find many comments with the **draft** keyword as a kind of tag.

They are still "**// gp** :" comments in the code to point out things that remain to be done.  

The little classes representing file content lines and inheriting the same interface were removed. They were more complicated and unclear than they should. So the **FileContent** class was changed as well. 

Reminder : 
- Getting the duration collector retrieve all data it's supposed
- duration collector retrieves data inside each MC year. In case of a study with several years, make sure that the number of calls in **time_measurement.txt** is not **1** but more.
- Check if there is a concurrence issue in filling the duration collector inside years played in parallel
- Take tags "**gp :**" comments into account 
-

guilpier-code committed:
Refactor info collect : make use of duration collector for printing few items in time_measurement_draft.txt

guilpier-code committed:
Refactor info collect : removing complicated and unclear classes representing lines of a file content

flomnes commented on src/libs/antares/benchmarking/info_collectors.h:
```diff
@@ -0,0 +1,86 @@
+#include <antares/study.h>
+#include "file_content.h"
+
+using namespace std;
```
We don't want to have this `using namespace` whenever we include the header.

guilpier-code replied:
Done

flomnes reviewed (commented):
1.  IMO, in the file contents, we should keep at least a key and value.

```cpp
struct FileContent {
struct KeyValue {
  const std::string key;
  const std::string value;
  std::write(std::ostream& os, const std::string& sep) {
    os << key << sep << value << '\n';
}
};
std::vector<KeyValue> lines_;
};
```
So that different formats can be used (CSV, JSON, etc.)

2. Remove existing data structures, only pass the new one to the Simulation object.
3. The number of calls should be in a different key-value pair
```txt
mc_years_time = 221
mc_years_nb_calls = 5
```

4. Add sections ?

```cpp
std::map<
std::string, // section name
std::vector<KeyValue> // key-value pairs
>
```

flomnes commented on src/libs/antares/benchmarking/info_collectors.cpp:
```diff
@@ -0,0 +1,158 @@
+			file_content.addDurationLine(name, (unsigned int)duration_sum, (int)durations.size());
+		}
+
+		file_content.addNameValueLine("total", (unsigned int)total_duration_);
```
I'm not sure we want to sum all durations in this way. There may be gaps in our time measurements, so this time measure will be inaccurate.

guilpier-code replied:
Done

guilpier-code committed:
Refactor info collect : adding duration measurement in the years job + measurement of hydro allocation

guilpier-code committed:
Refactor info collect : adding measurement of year by year export

guilpier-code committed:
Refactor info collect : adding measurement of all mc years, post-processing & TS generation

guilpier-code committed:
Refactor info collect : removing former way to collect durations in simulation (step 1)

guilpier-code committed:
Refactor info collect : removing former way to collect durations in simulation (step 2)

guilpier-code committed:
Refactor info collect : corrections due to review

- Remove storage and computation of total duration in DurationCollector class
- Remove an unwanted using namespace std in a header

guilpier-code committed:
Refactor info collect : introducing a null object instead of a duration collector

This design pattern is a kind of mock object that does nothing.
Useful to use the simulation object in isolation : the simulation needs a duration collector; in end-to-end tests we use a null duration collector
when building the simulation.

guilpier-code committed:
Refactor info collect : enhancing a mutex handling in the duration collector

guilpier-code committed:
[skip ci] Refactor info collect : changing file to be printed into an INI file (works, but needs enhancement)

- File content is now a map of maps : its can contains sections and inner sections parameters
- The printed file is now a ini file.

guilpier-code committed:
Refactor info collect : enhancements (renaming, move arguments from constructor to functions, ...)

flomnes commented on src/solver/main.cpp:
```diff
@@ -124,8 +124,7 @@ int main(int argc, char** argv)
 
-    application.writeElapsedTime();
-    application.writeStudyInfos();
+    application.writeInfoOnExecution();
```
Nice !

flomnes commented on src/libs/antares/benchmarking/info_collectors.h:
```diff
@@ -0,0 +1,100 @@
+    virtual void addDuration(std::string name, int64_t duration) = 0;
+};
+
+class NullDurationCollector : public IDurationCollector
+{
+public:
+    NullDurationCollector() = default;
+    void addDuration(std::string name, int64_t duration) override { /* Do nothing */ }
+};
```
I suggest moving this definition to **tests/end-to-end/simple-study.cpp** where it is used. 

Also, comment out `name` and `duration` to avoid "warning : unused parameter".

guilpier-code replied:
1. moving **NullDurationCollector** to end-to-end tests : if we do that, we separate **NullDurationCollector** from its class hierarchy and we can't easily reuse **NullDurationCollector** somewhere else. To be discussed.
2. Done

flomnes replied:
Do we really want to re-use this class elsewhere ?

flomnes commented on src/libs/antares/benchmarking/info_collectors.h:
```diff
@@ -0,0 +1,100 @@
+class IDurationCollector
+{
+public:
+    virtual void addDuration(std::string name, int64_t duration) = 0;
```
```suggestion
    virtual void addDuration(const std::string& name, int64_t duration) = 0;
```

guilpier-code replied:
Done

flomnes commented on src/libs/antares/benchmarking/file_content.cpp:
```diff
@@ -0,0 +1,32 @@
+
+	void FileContent::addDurationItem(std::string name, unsigned int duration, int nbCalls)
+	{
+		sections_["Durations"][name + "--duration"] = to_string(duration);
+		sections_["Durations"][name + "--nb_calls"] = to_string(nbCalls);
```
```suggestion
		addItemToSection("Durations", name + "_duration", duration);
		addItemToSection("Durations", name + "_nb_calls", nbCalls);
```

guilpier-code replied:
Done

flomnes commented on src/libs/antares/benchmarking/file_content.cpp:
```diff
@@ -0,0 +1,32 @@
+		return sections_.end();
+	}
+
+	void FileContent::addItemToSection(string section, string key, int value)
+	{
+		sections_[section][key] = to_string(value);
+	}
+
+	void FileContent::addItemToSection(string section, string key, string value)
+	{
+		sections_[section][key] = value;
+	}
+
+	void FileContent::addDurationItem(std::string name, unsigned int duration, int nbCalls)
```
1. Pass strings by const-ref
2. Use `string` version in the `int` implementation.

```suggestion
	void FileContent::addItemToSection(const std::string& section, const std::string& key, int value)
	{
		this->addItemToSection(section, key, std::to_string(value));
	}

	void FileContent::addItemToSection(const std::string& section, const std::string& key, const std::string& value)
	{
		sections_[section][key] = value;
	}

	void FileContent::addDurationItem(const std::string& name, unsigned int duration, int nbCalls)
```

guilpier-code replied:
Done

flomnes commented on src/libs/antares/benchmarking/file_content.h:
```diff
@@ -0,0 +1,31 @@
+#include <string>
+#include <map>
+
+#include <memory>
```
```suggestion
```

guilpier-code replied:
Done

flomnes commented on src/libs/antares/benchmarking/file_content.h:
```diff
@@ -0,0 +1,31 @@
+
+private:
+    // Data of the file content
+    std::map<std::string, std::map<std::string, std::string>> sections_;
```
```suggestion
    std::map<std::string, // Section name
    std::map<std::string, std::string> // Key-values
    > sections_;
```

guilpier-code replied:
Done

flomnes commented on src/libs/antares/benchmarking/file_content.h:
```diff
@@ -0,0 +1,31 @@
+    iterator endSections();
+
+    void addItemToSection(std::string section, std::string key, int value);
+    void addItemToSection(std::string section, std::string key, std::string value);
```
Pass `std::strings` by const ref

guilpier-code replied:
Done

flomnes commented on src/libs/antares/benchmarking/file_writer.cpp:
```diff
@@ -0,0 +1,32 @@
+	{
+		Antares::IniFile ini;
+
+		FileContent::iterator it_section = fileContent_.firstSection();
+		for (; it_section != fileContent_.endSections(); it_section++)
+		{
+			Antares::IniFile::Section* section = ini.addSection(it_section->first);
+			for (pair<string, string> line : it_section->second)
+				section->add(line.first, line.second);
+		}
```
Use `auto` whenever possible ?

guilpier-code replied:
Done

flomnes commented on src/solver/application.cpp:
```diff
@@ -525,32 +525,30 @@ void Application::readDataForTheStudy(Data::StudyLoadOptions& options)
-    filePath.clear() << pStudy->folderOutput << Yuni::IO::Separator << "study-info.txt";
-    Benchmarking::StudyInfoCSVwriter study_info_writer(filePath, study_info);
-    study_info_writer.flush();
+    Yuni::String filePath;
+    filePath.clear() << pStudy->folderOutput << Yuni::IO::Separator << "execution_info.ini";
+    Benchmarking::iniFilewriter ini_file_writer(filePath, file_content);
+    ini_file_writer.flush();
```
The interface class `FileWriter` isn't very useful for now since `iniFileWriter` is its only child, and there is no need for dynamic polymorphism. Remove it ?

guilpier-code replied:
If we are sure we want an ini file, we can remove it. But :
- We might want a Json file eventually
- Like this, we separate the essence of a writer from a particular implementation.

To be discussed

flomnes replied:
Ok ! Let's keep it but then, make sure to use dynamic polymorphism in the application. Something like
```cpp
// application.h
std::unique_ptr<FileWriter> writer;
// application.cpp
writer = std::make_unique<iniFileWriter>(/* args */);
```

Also, I suggest renaming `FileWriter` to `IFileWriter`

flomnes commented on src/solver/application.h:
```diff
@@ -50,8 +50,7 @@ class Application : public Yuni::IEventObserver<Application, Yuni::Policy::Singl
 
-    void writeElapsedTime();
-    void writeStudyInfos();
+    void writeInfoOnExecution();
```
writeExectutionInfo

guilpier-code replied:
Done

flomnes commented on src/libs/antares/benchmarking/info_collectors.h:
```diff
@@ -0,0 +1,100 @@
+    const Antares::Data::Study& study_;
+
+    // ----------------------------------------------------------------------------------------
+    // TODO : add some more pieces of information about study : see following data members.
```
Do we want to add more info ? Or is this an old TODO note ?

guilpier-code replied:
We wanted more pieces of information : they are pointed by the unused data members below the comment.
Do we still want them ?

flomnes replied:
No, you may remove these data members.

guilpier-code replied:
Done

flomnes commented on src/libs/antares/benchmarking/info_collectors.cpp:
```diff
@@ -0,0 +1,152 @@
+	// -------------------------------------
+	void DurationCollector::toFileContent(FileContent& file_content)
+	{		
+		for (pair<string, vector<int64_t>> element : duration_items_)
+		{
+			std::string name = element.first;
+			vector<int64_t> durations = element.second;
+			int64_t duration_sum = 0;
+			for (auto& duration : durations)
+			{
+				duration_sum += duration;
+			}
```
Use `auto&` to avoid unnecessary copies. Also, you could keep only the sum of durations, removing the need for loop at lines 137-140.

guilpier-code replied:
Using **auto&** to avoid unnecessary copies : done
For the other point : I' m not sure to understand your remark. If you mean we don't need vectors as value of map **duration_items_** (of type `map<string, vector<int64_t>>`), but **duration_items_** should be defined as a `map<string, unsigned int>` to relate a name to a duration sum, I disagree : we need the vector size below the loop.

I replaced the loop with an accumulation though, to make code more concise.

flomnes replied:
I understand. But if we wanted to know the number of calls, we could just increment a counter. On the other hand, it doesn't really matter since we don't expect large vectors, so memory usage won't be an issue.

flomnes commented on src/libs/antares/benchmarking/info_collectors.cpp:
```diff
@@ -0,0 +1,152 @@
+	void DurationCollector::addDuration(std::string name, int64_t duration)
+	{	
+		const std::lock_guard<std::mutex> lock(mutex_);
+		duration_items_[name].push_back(duration);
```
```suggestion
		duration_items_[name] += duration;
```

guilpier-code replied:
See remark above

flomnes commented on src/libs/antares/benchmarking/info_collectors.h:
```diff
@@ -0,0 +1,100 @@
+    void addDuration(std::string name, int64_t duration) override;
+
+private:
+    map<string, vector<int64_t>> duration_items_;
```
```suggestion
    map<string, int64_t> duration_items_;
```
There is no need to keep all durations, only their sum.

guilpier-code replied:
See remark above

flomnes commented on src/libs/antares/benchmarking/info_collectors.cpp:
```diff
@@ -0,0 +1,152 @@
+	// ---------------------------------
+	void SimulationInfoCollector::toFileContent(FileContent& file_content)
+	{
+		file_content.addItemToSection("from simulation", "variables", opt_info_.nbVariables);
+		file_content.addItemToSection("from simulation", "constraints", opt_info_.nbConstraints);
+		file_content.addItemToSection("from simulation", "non-zero coefficients", opt_info_.nbNonZeroCoeffs);
```
Section "optimization problem"

guilpier-code replied:
Done

flomnes commented on src/libs/antares/benchmarking/info_collectors.cpp:
```diff
@@ -0,0 +1,152 @@
+
+	void StudyInfoCollector::areasCountToFileContent(FileContent& file_content)
+	{
+		file_content.addItemToSection("from study", "areas", study_.areas.size());
```
Section "study"

guilpier-code replied:
Done

flomnes commented on src/solver/application.cpp:
```diff
@@ -421,14 +419,16 @@ void Application::readDataForTheStudy(Data::StudyLoadOptions& options)
-    loadTimer.stop();
+
+    timer.stop();
+    pDurationCollector.addDuration("study_loading", timer.get_duration());
```
[Style] You could pass `timer` as the 2nd argument, and call `get_duration` inside `addDuration`

guilpier-code replied:
We could, but the code is clear as it is now. If we pass the `Timer` in the **addDuration** method and call stop() inside that method, it's shorter but less clear I think.

flomnes replied:
We can keep the call to `Timer::stop` outside, but place the call to `Timer::get_duration` inside `addDuration`

flomnes reviewed (commented):
I wonder if in some cases the simulation may get slowed down by the `mutex`, when executing years in parallel. I'm talking about one thread waiting for the other, etc.

Also, we need to make sure that there is one single instance of `mutex`, not one per thread. Otherwise the lock mechanism won't lock.

guilpier-code committed:
[skip ci] Refactor info collect : some cleaning

guilpier-code committed:
Refactor info collect : corrections due to review

guilpier-code committed:
Refactor info collect : cleaning due to review
